### PR TITLE
Extend `beluga::estimate` for SE3

### PR DESCRIPTION
**Pull request by serraramiro1**
### Proposed changes

Exactly what the title states. 

#### Type of change

- [ ] 🐛 Bugfix (change which fixes an issue)
- [x] 🚀 Feature (change which adds functionality)
- [ ] 📚 Documentation (change which fixes or extends documentation)

### Checklist

_Put an `x` in the boxes that apply. This is simply a reminder of what we will require before merging your code._

- [x] Lint and unit tests (if any) pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] All commits have been signed for [DCO](https://developercertificate.org/)

### CODE REVIEW

File: beluga/include/beluga/algorithm/estimation.hpp [hunk 1]
```diff
@@ -102,6 +104,68 @@ Sophus::Matrix2<Scalar> calculate_covariance(Range&& range, const Sophus::Vector
...
+ */
+template <
+    class Poses,
+    class Weights,
+    class Pose = ranges::range_value_t<Poses>,
+    class Scalar = typename Pose::Scalar,
+    typename = std::enable_if_t<std::is_same_v<Pose, typename Sophus::SE3<Scalar>>>>
+std::pair<Sophus::SE3<Scalar>, Sophus::Matrix6<Scalar>> estimate(Poses&& poses, Weights&& weights) {
+  assert(std::size(poses) == std::size(weights));
+  assert(std::size(poses) > 0);
```
Comment (nahueespinosa) [inline beluga/include/beluga/algorithm/estimation.hpp:131]: We can add `ASSERT_DEBUG_DEATH` tests for these (requires `#include <gtest/gtest-death-test.h>`).

File: beluga/include/beluga/algorithm/estimation.hpp [hunk 2]
```diff
@@ -102,6 +104,68 @@ Sophus::Matrix2<Scalar> calculate_covariance(Range&& range, const Sophus::Vector
...
+  auto translation_view = poses | ranges::views::transform([](const auto& pose) { return pose.translation(); });
```
Comment (nahueespinosa) [inline beluga/include/beluga/algorithm/estimation.hpp:132]: Unused?

File: beluga/include/beluga/algorithm/estimation.hpp [hunk 3]
```diff
@@ -102,6 +104,68 @@ Sophus::Matrix2<Scalar> calculate_covariance(Range&& range, const Sophus::Vector
...
+  auto poses_view = poses | ranges::views::common;
+  auto weights_view = weights | ranges::views::common;
+  const auto weights_sum = std::accumulate(weights_view.begin(), weights_view.end(), 0.0);
+  auto normalized_weights_view =
+      weights_view | ranges::views::transform([weights_sum](const auto& weight) { return weight / weights_sum; });
+
+  const auto weighted_translation = [](const Pose& pose, const auto& weight) { return pose.translation() * weight; };
```
Comment (nahueespinosa) [inline beluga/include/beluga/algorithm/estimation.hpp:139]:
```suggestion
  const auto weighted_translation = [](const Pose& pose, auto weight) { return pose.translation() * weight; };
```

File: beluga/include/beluga/algorithm/estimation.hpp [hunk 4]
```diff
@@ -102,6 +104,68 @@ Sophus::Matrix2<Scalar> calculate_covariance(Range&& range, const Sophus::Vector
...
+
+  const auto weighted_rotation_vector = [](const Pose& pose, const auto& weight) {
+    return (pose.so3().log() * weight).eval();
+  };
+
+  const Sophus::Vector3<Scalar> mean_rotation_vector = std::transform_reduce(
+      poses_view.begin(), poses_view.end(), normalized_weights_view.begin(), Sophus::Vector3<Scalar>::Zero().eval(),
+      std::plus{}, weighted_rotation_vector);
```
Comment (nahueespinosa) [inline beluga/include/beluga/algorithm/estimation.hpp:147]: Consider using the lambda directly in the function call:

```suggestion
  const Sophus::Vector3<Scalar> mean_rotation_vector = std::transform_reduce(
      poses_view.begin(), //
      poses_view.end(), //
      normalized_weights_view.begin(), //
      Sophus::Vector3<Scalar>::Zero().eval(), //
      std::plus{}, //
      [](const Pose& pose, auto weight) { return (pose.so3().log() * weight).eval(); });
```

Same elsewhere.

File: beluga/test/beluga/algorithm/test_estimation.cpp [hunk 5]
```diff
@@ -252,5 +263,37 @@ TEST_F(ScalarEstimation, NonUniformWeightOverload) {
...
 }
+TEST_F(PoseCovarianceEstimation, MultiVariateNormalSE3) {
+  constexpr double kTolerance = 0.01;
+  const auto expected_mean =
+      Sophus::SE3d{Sophus::SO3d::exp(Eigen::Vector3d{-0.17, 0.25, 0.1}), Eigen::Vector3d{1.0, 2.0, 3.0}};
+  const Eigen::Matrix<double, 6, 6> expected_cov = Eigen::Matrix<double, 6, 6>::Identity() * 2e-1;
+  auto distribution = beluga::MultivariateNormalDistribution{expected_mean, expected_cov};
+
+  const auto samples =
+      beluga::views::sample(distribution) | ranges::views::take_exactly(500'000) | ranges::to<std::vector>;
```
Comment (nahueespinosa) [inline beluga/test/beluga/algorithm/test_estimation.cpp:274]: nit:
```suggestion
  const auto samples = //
      beluga::views::sample(distribution) | //
      ranges::views::take_exactly(500'000) | //
      ranges::to<std::vector>;
```

File: beluga/test/beluga/algorithm/test_estimation.cpp [hunk 6]
```diff
@@ -252,5 +263,37 @@ TEST_F(ScalarEstimation, NonUniformWeightOverload) {
...
+  const auto [mean, cov] =
+      beluga::estimate(samples, ranges::views::repeat_n(1.0, static_cast<std::ptrdiff_t>(samples.size())));
```
Comment (nahueespinosa) [inline beluga/test/beluga/algorithm/test_estimation.cpp:276]: Can we extend this to 3D?
https://github.com/Ekumen-OS/beluga/blob/cc6b52c5412612ebed74b9d10509179913d5e62d/beluga/include/beluga/algorithm/estimation.hpp#L281-L306

**Review by nahueespinosa — COMMENTED** [review]
@serraramiro1 Awesome!

File: beluga/test/beluga/algorithm/test_estimation.cpp [hunk 7]
```diff
@@ -252,5 +263,37 @@ TEST_F(ScalarEstimation, NonUniformWeightOverload) {
   ASSERT_NEAR(mean, 4.300, kTolerance);
   ASSERT_NEAR(standard_deviation, 2.026, kTolerance);
 }
+TEST_F(PoseCovarianceEstimation, MultiVariateNormalSE3) {
```
Comment (nahueespinosa) [inline beluga/test/beluga/algorithm/test_estimation.cpp:266]: nit: Add newline.

File: beluga/test/beluga/algorithm/test_estimation.cpp [hunk 8]
```diff
@@ -252,5 +263,37 @@ TEST_F(ScalarEstimation, NonUniformWeightOverload) {
...
+  constexpr double kTolerance = 0.01;
+  const auto expected_mean =
+      Sophus::SE3d{Sophus::SO3d::exp(Eigen::Vector3d{-0.17, 0.25, 0.1}), Eigen::Vector3d{1.0, 2.0, 3.0}};
```
[the remark this answers is not in the document]
Comment (serraramiro1) [reply beluga/test/beluga/algorithm/test_estimation.cpp:272]: Why not?
This is just an arbitrary 3D rotation

**Review by nahueespinosa — APPROVED** [review]
LGTM!

Having @hidmic's opinion would be nice.

**Comment by serraramiro1** [comment]
@hidmic You've been invoked! :mage:

File: beluga/include/beluga/algorithm/estimation.hpp [hunk 9]
```diff
@@ -102,6 +104,72 @@ Sophus::Matrix2<Scalar> calculate_covariance(Range&& range, const Sophus::Vector
...
+      weights_view |  //
+      ranges::views::transform([weights_sum](const auto weight) { return weight / weights_sum; });
+
+  const Sophus::Vector3<Scalar> mean_rotation_vector = std::transform_reduce(
+      poses_view.begin(),                      //
+      poses_view.end(),                        //
+      normalized_weights_view.begin(),         //
+      Sophus::Vector3<Scalar>::Zero().eval(),  //
+      std::plus{},                             //
+      [](const Pose& pose, const auto& weight) { return (pose.so3().log() * weight).eval(); });
```
[the remark this answers is not in the document]
Comment (serraramiro1) [reply beluga/include/beluga/algorithm/estimation.hpp:145]: I just found out about https://github.com/strasdat/Sophus/blob/main/sophus/average.hpp , so I'll use that instead and trust Sophus on this one.

File: beluga/include/beluga/algorithm/estimation.hpp [hunk 10]
```diff
@@ -102,6 +104,72 @@ Sophus::Matrix2<Scalar> calculate_covariance(Range&& range, const Sophus::Vector
...
+      poses_view.begin(),                          //
+      poses_view.end(),                            //
+      normalized_weights_view.begin(),             //
+      Eigen::Matrix<Scalar, 6, 6>::Zero().eval(),  //
+      std::plus{},                                 //
+      [inverse_mean = mean.inverse()](const Pose& pose, const auto weight) {
+        // Compute deviation from mean in Lie algebra (logarithm of SE3)
+        const Pose delta = inverse_mean * pose;
+        // Accumulate weighted covariance
+        return Eigen::Matrix<Scalar, 6, 6>{weight * (delta.log() * delta.log().transpose())};
```
[the remark this answers is not in the document]
Comment (serraramiro1) [reply beluga/include/beluga/algorithm/estimation.hpp:167]: Why do we want to change parametrizations here? We're already using this parametrization for the `MultivariateNormalDistribution`, so it makes sense for consistency purposes to keep the same parametrization here.
We can always add a conversion layer when appropriate.

Are you spotting a mathematical error here?

File: beluga/include/beluga/algorithm/estimation.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (hidmic) [reply beluga/include/beluga/algorithm/estimation.hpp:145]: Hmm. I like that it is bi-invariant but I don't like that it again only works in a close neighborhood. Worth checking what results we get if we average SE(3) directly vs. averaging $SO(3) \times R^3$ (i.e. using the `average` overload for SO(3), which I see works with quaternions :rocket:).

File: beluga/include/beluga/algorithm/estimation.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (hidmic) [reply beluga/include/beluga/algorithm/estimation.hpp:167]: Fair point. Came in with the ROS chip. +1 to consistency within Beluga.

`beluga_ros` will need to perform the conversion though ($R^{6\times6}$ covariances in [ROS messaging](https://docs.ros.org/en/noetic/api/geometry_msgs/html/msg/PoseWithCovariance.html) use a roll-pitch-yaw parametrization).

[the remark this answers is not in the document]
Comment (hidmic) [reply beluga/include/beluga/algorithm/estimation.hpp:167]: On a second look, we are using a biased estimator for sample variance. We are missing the Bessel correction. Check the corresponding section on [Wikipedia](https://en.wikipedia.org/wiki/Weighted_arithmetic_mean).

[the remark this answers is not in the document]
Comment (serraramiro1) [reply beluga/include/beluga/algorithm/estimation.hpp]: Made the modifications as per our offline discussion,  replicating Sophus' approach for quaternion averaging, decoupling rotation and translation for the mean computation, and adding the Bessel correction, LMK what you think.

File: beluga/include/beluga/algorithm/estimation.hpp [hunk 11]
```diff
@@ -35,6 +39,61 @@
...
+      max_eigenvalue = es.eigenvalues()[i];
+      max_eigenvector = es.eigenvectors().col(i);
+    }
+  }
+  Eigen::Quaternion<Scalar> quat;
+  quat.coeffs() <<                //
+      max_eigenvector[0].real(),  //
+      max_eigenvector[1].real(),  //
+      max_eigenvector[2].real(),  //
+      max_eigenvector[3].real();
```
Comment (hidmic) [inline beluga/include/beluga/algorithm/estimation.hpp:92]: @serraramiro1 nit: I *think* `Eigen::Quaternion<Scalar>{max_eigenvector.real()}` should get you the same result.

File: beluga/include/beluga/algorithm/estimation.hpp [hunk 12]
```diff
@@ -35,6 +39,61 @@
...
+ */
+template <
+    class QuaternionRange,
+    class WeightsRange,
+    class Scalar = typename ranges::range_value_t<QuaternionRange>::Scalar>
+Eigen::Quaternion<Scalar> weighted_average_quaternion(
+    const QuaternionRange& quaternions,
+    const WeightsRange& normalized_weights) {
+  // Implementation is based on https://github.com/strasdat/Sophus/blob/main/sophus/average.hpp#L133 with the variant of
+  // non-uniform weights.
```
Comment (hidmic) [inline beluga/include/beluga/algorithm/estimation.hpp:64]: @serraramiro1 nit: consider adding a reference to the report / paper we know this is coming from, for future reference.

File: beluga/include/beluga/algorithm/estimation.hpp [hunk 13]
```diff
@@ -35,6 +39,61 @@
...
+ * \brief Given a range with quaternions and a normalized range of weights, compute the weighted average quaternion.
+ *
+ * \tparam QuaternionRange A [sized range](https://en.cppreference.com/w/cpp/ranges/sized_range) type whose
+ *  value type is `Eigen::Quaternion<Scalar>`.
+ * \tparam WeightsRange A [sized range](https://en.cppreference.com/w/cpp/ranges/sized_range) type whose
+ *  value type is `Scalar`.
+ * \tparam Scalar The scalar type, matching those of the weights.
+ * \param quaternions Range of quaternions to compute the weighted average from.
+ * \param normalized_weights Range of normalized weights. With matching indices wrt \c quaternions . Non-normalized
+ * weights will yield incorrect results. \return The average quaternion.
```
Comment (hidmic) [inline beluga/include/beluga/algorithm/estimation.hpp:54]: @serraramiro1 nit: 
```suggestion
 * weights will yield incorrect results. 
 \return The average quaternion.
```

File: beluga/include/beluga/algorithm/estimation.hpp [hunk 14]
```diff
@@ -102,6 +161,79 @@ Sophus::Matrix2<Scalar> calculate_covariance(Range&& range, const Sophus::Vector
       mean);
 }
 
+/// Returns a pair consisting of the estimated mean pose and its covariance.
+/**
+ * Given a range of poses, computes the estimated pose by averaging the translation
+ * and rotation parts.
+ * Computes the covariance matrix of the translation and rotation parts to create a 6x6 covariance matrix.
+ * NOTE: The covariance matrix is in se3, the tangent space of the SE3 manifold. Users may perform
+ * the appropriate conversions to get the covariance matrix into their parametrization of interest.
```
Comment (hidmic) [inline beluga/include/beluga/algorithm/estimation.hpp:170]: @serraramiro1 consider expanding a bit, explaining the rationale for the choice, plus a reference to relevant literature.

File: beluga/test/beluga/algorithm/test_estimation.cpp [hunk 15]
```diff
@@ -253,4 +268,84 @@ TEST_F(ScalarEstimation, NonUniformWeightOverload) {
...
 }
 
+TEST_F(PoseCovarianceEstimation, MultiVariateNormalSE3) {
+  constexpr double kTolerance = 0.01;
+  const auto expected_mean =
+      Sophus::SE3d{Sophus::SO3d::exp(Eigen::Vector3d{-0.17, 0.25, 0.1}), Eigen::Vector3d{1.0, 2.0, 3.0}};
+  const Eigen::Matrix<double, 6, 6> expected_cov = Eigen::Matrix<double, 6, 6>::Identity() * 2e-1;
+  auto distribution = beluga::MultivariateNormalDistribution{expected_mean, expected_cov};
+  const auto samples = beluga::views::sample(distribution) |   //
+                       ranges::views::take_exactly(500'000) |  //
```
Comment (hidmic) [inline beluga/test/beluga/algorithm/test_estimation.cpp:278]: @serraramiro1 consider using fewer samples, `100'000` should give you a good enough approximation and will improve test runtime for debug builds.

**Review by hidmic — COMMENTED** [review]
Second pass, looking great.

File: beluga/include/beluga/algorithm/estimation.hpp [hunk 16]
```diff
@@ -35,6 +39,61 @@
...
+    const QuaternionRange& quaternions,
+    const WeightsRange& normalized_weights) {
+  // Implementation is based on https://github.com/strasdat/Sophus/blob/main/sophus/average.hpp#L133 with the variant of
+  // non-uniform weights.
+  const size_t num_quaternions = std::size(quaternions);
+  assert(num_quaternions >= 1);
+  Eigen::Matrix<Scalar, 4, Eigen::Dynamic> all_quaternions(4, num_quaternions);
+
+  int i = 0;
+  for (const auto& [quaternion, normalized_weight] : ranges::views::zip(quaternions, normalized_weights)) {
```
[the remark this answers is not in the document]
Comment (serraramiro1) [reply beluga/include/beluga/algorithm/estimation.hpp]: Nice one, done!

File: beluga/include/beluga/algorithm/estimation.hpp [hunk 17]
```diff
@@ -35,6 +39,61 @@
...
+
+  std::complex<Scalar> max_eigenvalue = es.eigenvalues()[0];
+  Eigen::Matrix<std::complex<Scalar>, 4, 1> max_eigenvector = es.eigenvectors().col(0);
+
+  for (int i = 1; i < 4; i++) {
+    if (std::norm(es.eigenvalues()[i]) > std::norm(max_eigenvalue)) {
+      max_eigenvalue = es.eigenvalues()[i];
+      max_eigenvector = es.eigenvectors().col(i);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (serraramiro1) [reply beluga/include/beluga/algorithm/estimation.hpp]: Done

File: beluga/include/beluga/algorithm/estimation.hpp (hunk 11 above)
Comment (serraramiro1) [reply beluga/include/beluga/algorithm/estimation.hpp]: Not quite, added a coment

File: beluga/include/beluga/algorithm/estimation.hpp (hunk 12 above)
Comment (serraramiro1) [reply beluga/include/beluga/algorithm/estimation.hpp:67]: Done

**Review by hidmic — APPROVED** [review]
LGTM with green CI

**Pull request merged**